benches/hvm_cpu.rs

```diff
@@ -29,57 +35,75 @@ fn load_from_lang(file: &str, size: usize, replace: Option<(&str, &str)>) -> (ru
-      b.iter_batched(|| net.clone(), |mut net| black_box(net.normal(&book)), criterion::BatchSize::SmallInput);
-    });
-    // Church exponentiation
-    // n=24 uses at most 1107 elements in heap(1 << 12 = 4096)
-    let (book, net) = load_from_core("./benches/programs/church_exp.hvmc", 1 << 12, Some(("{n}", &n.to_string())));
-    group.bench_with_input(criterion::BenchmarkId::new("exponentiation", n), &n, |b, &_n| {
-      b.iter_batched(|| net.clone(), |mut net| black_box(net.normal(&book)), criterion::BatchSize::SmallInput);
-    });
-  }
+fn extract_size(code: &str) -> (usize, &str) {
+  code
+    .strip_prefix("// size = ")
+    .and_then(|code| code.split_once('\n'))
+    .and_then(|(size, rest)| {
+      match size.split_ascii_whitespace().collect::<Vec<_>>().as_slice() {
+        [a, "<<", b] => a.parse::<usize>().ok().zip(b.parse::<usize>().ok()).map(|(a, b)| a << b),
+        [a] => a.parse().ok(),
+        _ => None,
+      }
+      .map(|size| (size, rest))
+    })
+    .expect("failed to extract bench size")
 }
 
-fn tree_benchmark(c: &mut Criterion) {
-  let mut group = c.benchmark_group("tree");
-  // Allocates a big tree
-  for n in [12] {
-    let (book, net) = load_from_core("./benches/programs/alloc_big_tree.hvmc", 16 << n, Some(("{n}", &n.to_string())));
-    group.bench_with_input(criterion::BenchmarkId::new("allocation", n), &n, |b, &_n| {
-      b.iter_batched(|| net.clone(), |mut net| black_box(net.normal(&book)), criterion::BatchSize::SmallInput);
-    });
-  }
+fn run_programs_dir(c: &mut Criterion) {
+  let root = PathBuf::from(format!("{}/benches/programs", env!("CARGO_MANIFEST_DIR")));
+  run_dir(&root, None, c);
 }
 
-fn binary_counter_benchmark(c: &mut Criterion) {
-  let mut group = c.benchmark_group("binary-counter");
-  // Decrements a BitString until it is zero
-  for n in [12] {
-    let (book, net) = load_from_core("./benches/programs/dec_bits.hvmc", 16 << n, Some(("{n}", &n.to_string())));
-    group.bench_with_input(criterion::BenchmarkId::new("single", n), &n, |b, &_n| {
-      b.iter_batched(|| net.clone(), |mut net| black_box(net.normal(&book)), criterion::BatchSize::SmallInput);
-    });
-  }
-  // Decrements 2^N BitStrings until they reach zero (ex3)
-  for n in [6] {
-    let (book, net) = load_from_core("./benches/programs/dec_bits_tree.hvmc", 128 << n, Some(("{n}", &n.to_string())));
-    group.bench_with_input(criterion::BenchmarkId::new("many", n), &n, |b, &_n| {
-      b.iter_batched(|| net.clone(), |mut net| black_box(net.normal(&book)), criterion::BatchSize::SmallInput);
-    });
+fn run_dir(dir: &PathBuf, group: Option<&str>, c: &mut Criterion) {
+  let files = std::fs::read_dir(dir).unwrap();
+
+  for file in files.flatten() {
+    let file_path = &file.path();
+    let file_name = file_path.file_stem().unwrap().to_string_lossy();
+
+    let Some(ext) = file_path.extension() else {
+      if file_path.is_dir() {
+        match group {
```

Comment:
this didn't make a lot of sense, I thought `run_dir` is a recursive function, that would take the folder's name as the group's name, can you explain it in more detail?

---

- programs
  - church
    - mul
    - exp
  - neg_fusion

would result in(`<GROUP>/<TEST> | <TEST>`):
- church/mul
- church/exp
- neg_fusion

Reply:
Yes, that was the intent as I guessed the grouping could affect how the benchmarks results are presented. Should we just have the file name with no folder/structure information?

Reply:
![image](https://github.com/HigherOrderCO/hvm-core/assets/19491161/f3849968-48a6-4c09-a11a-be1e2a6987cc)
Confirmed in the generated index, I did it this way to maintain the structure on the report